### PR TITLE
Nf-test

### LINKED ISSUE #91: Implement tests with nf-test
Pytest workflow makes it easier to maintain larger test suites and also allows to check outputs. 

cf. https://github.com/nf-core/tools/issues/605, https://github.com/nf-core/rnaseq/pull/546, https://github.com/nf-core/sarek/tree/dev/tests

Testing scenarios: 
 * Run pipeline with more than one file
 * different parameter combinations
 * different filenames and folder structures for cellranger
 * run subworkflows with and without index
 * different protocols, if we manage to get the test files

### PR DESCRIPTION
**Pull request by grst**
Close #91 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/scrnaseq/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/scrnaseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by grst** [comment]
Thanks, I don't see any major issue... Could probably add more/all output files to the snapshot.

Asking for additional review by @sateeshperi who has worked more with nf-test afaik

**Comment by fmalmeida** [comment]
Cool. I will then first work in prototyping the tests for the other aligners following the same structure, and in the end, based on review and all we can check if we add or remove some files.

😄

**Comment by fmalmeida** [comment]
Hi @grst ,
I have added the nf-test run for the main aligners, alevin, star, cellranger and kallisto.
Thus, what is required now is to change the github actions to run them instead of pytests and to have a proper review discussion over the testings.

File: tests/main_pipeline_alevin.test [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            //
+            // Check if files were produced
+            //
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_X_alevin_results/af_quant/alevin/quants_mat_cols.txt" ).exists()
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_X_alevin_results/af_quant/alevin/quants_mat.mtx" ).exists()
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_X_alevin_results/af_quant/alevin/quants_mat_rows.txt" ).exists()
+
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_Y_alevin_results/af_quant/alevin/quants_mat_cols.txt" ).exists()
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_Y_alevin_results/af_quant/alevin/quants_mat.mtx" ).exists()
+            assert new File( "${outputDir}/results_alevin/alevin/Sample_Y_alevin_results/af_quant/alevin/quants_mat_rows.txt" ).exists()
```
Comment (grst) [inline tests/main_pipeline_alevin.test:50]: Do we need this if the same files are already part of the snapshot?

**Review by grst — COMMENTED** [review]
Really cool, thanks! 
I think you could steal the CI part from e.g. [fetchngs](https://github.com/nf-core/fetchngs/blob/master/.github/workflows/ci.yml). Or maybe someone in the `nf-test` channel on slack has better advice. 

I'm happy to merge this soon, this'll add a lot of value!

File: tests/main_pipeline_alevin.test (hunk 1 above)
Comment (fmalmeida) [reply tests/main_pipeline_alevin.test:50]: Nope. Not needed.
This is a "footprint" from the conversion which I forgot to remove. I will take it out from all the tests where it happens.

Comment (fmalmeida) [reply tests/main_pipeline_alevin.test]: Just fixed this one.

**Comment by fmalmeida** [comment]
Hi @grst ,
Have addressed your comment and also changed the `ci.yml` to use nf-test. I think it is working.

**Review by grst — COMMENTED** [review]
lol, I can't approve because I opened the PR but didn't make a single commit :sweat_smile: 

LGTM though - you can probably approve yourself :D I also asked for another review in the nf-core slack.

**Comment by fmalmeida** [comment]
Let's see if someone else reviews it then. If not, I can approve tomorrow.

File: tests/main_pipeline_alevin.test [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        then {
+
+            //
+            // General assertions
+            //
+
+            // Did it finish successfully?
+            assert workflow.success
```
Comment (sateeshperi) [inline tests/main_pipeline_alevin.test:28]: recommend wrapping all assertions in `assertAll()` block....rest LGTM

Comment (fmalmeida) [reply tests/main_pipeline_alevin.test]: Done.

**Review by fmalmeida — APPROVED** [review]
Since all was resolved, I am approving it.

**Pull request merged**